src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/PropertyPage.cs

```diff
@@ -0,0 +1,404 @@
+    {
+        private IPropertyPageSite _site = null;
+        private bool _isDirty = false;
+        private bool _ignoreEvents = false;
+        private bool _useJoinableTaskFactory = true;
+        private IVsDebugger _debugger;
+        private uint _debuggerCookie;
+
+        // WIN32 Constants
+        private const int
+            WM_KEYFIRST = 0x0100,
+            WM_KEYLAST = 0x0108,
+            WM_MOUSEFIRST = 0x0200,
+            WM_MOUSELAST = 0x020A,
+            SW_HIDE = 0;
+
+        internal static class NativeMethods
+        {
+            public const int
+                S_OK = 0x00000000;
+        }
+
+        protected abstract string PropertyPageName { get; }
+
+        internal PropertyPage()
+        {
+            AutoScroll = false;
+        }
+
+        // For unit testing
+        internal PropertyPage(bool useJoinableTaskFactory)
+        {
+            _useJoinableTaskFactory = useJoinableTaskFactory;
+        }
+
+        internal UnconfiguredProject _unconfiguredProject { get; set; }
+        internal IProjectThreadingService _threadHandling { get; set; }
+        
+        ///--------------------------------------------------------------------------------------------
+        /// <summary>
+        /// Property. Gets or sets whether the page is dirty. Dirty status is pushed to owner property sheet
+        /// </summary>
+        ///--------------------------------------------------------------------------------------------
+        protected bool IsDirty
+        {
+            get { return _isDirty; }
+            set
+            {
+                // Only process real changes
+                if (value != _isDirty && !_ignoreEvents)
+                {
+                    _isDirty = value;
+                    // If dirty, this causes Apply to be called
+                    if (_site != null)
+                        _site.OnStatusChange((uint)(_isDirty ? PROPPAGESTATUS.PROPPAGESTATUS_DIRTY : PROPPAGESTATUS.PROPPAGESTATUS_CLEAN));
+                }
+            }
+        }
+
+        public List<IVsBrowseObjectContext> ContextObjects { get; private set; }
+
+        /// <summary>
+        /// Helper to wait on async tasks
+        /// </summary>
+        private T WaitForAsync<T>(Func<Task<T>> asyncFunc)
+        {
+            if (!_useJoinableTaskFactory)
+            {
+                // internal test usage
+                Task<T> t = asyncFunc();
+                return t.Result;
+            }
+            //Assert(_threadHandling != null);
+            return _threadHandling.ExecuteSynchronously<T>(asyncFunc);
+        }
+
+        /// <summary>
+        /// Helper to wait on async tasks
+        /// </summary>
+        private void WaitForAsync(Func<Task> asyncFunc)
+        {
+            // Real VS execution
+            if (!_useJoinableTaskFactory)
+            {
+                // internal test usage
+                asyncFunc().Wait();
+                return;
+            }
```

Comment:
We should remove stuff like this. The thread handling mock I checked in handles this correctly.